Update encoding_util.go

liuzheng opened the pull request:
fix this panic
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x96df96]

goroutine 60 [running]:
github.com/amitbet/vnc2video.(*VncCanvas).RemoveCursor(0x0, 0xc000139b00, 0xc0003faedf)
        D:/go1.13.8.windows-amd64/gopath/pkg/mod/github.com/amitbet/vnc2video@v0.0.0-20190616012314-9d50b9dab1d9/encoding_util.go:61 +0x26
github.com/amitbet/vnc2video.(*DefaultClientMessageHandler).Handle.func2(0xc0003434d0, 0xbce180, 0xc000120a00, 0xc0002e8b20, 0xc00011e500, 0xc000139b00)
        D:/go1.13.8.windows-amd64/gopath/pkg/mod/github.com/amitbet/vnc2video@v0.0.0-20190616012314-9d50b9dab1d9/client.go:303 +0x2ab
created by github.com/amitbet/vnc2video.(*DefaultClientMessageHandler).Handle
        D:/go1.13.8.windows-amd64/gopath/pkg/mod/github.com/amitbet/vnc2video@v0.0.0-20190616012314-9d50b9dab1d9/client.go:285 +0x2b8
```

liuzheng committed:
Update encoding_util.go

fix this panic
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x96df96]

goroutine 60 [running]:
github.com/amitbet/vnc2video.(*VncCanvas).RemoveCursor(0x0, 0xc000139b00, 0xc0003faedf)
        D:/go1.13.8.windows-amd64/gopath/pkg/mod/github.com/amitbet/vnc2video@v0.0.0-20190616012314-9d50b9dab1d9/encoding_util.go:61 +0x26
github.com/amitbet/vnc2video.(*DefaultClientMessageHandler).Handle.func2(0xc0003434d0, 0xbce180, 0xc000120a00, 0xc0002e8b20, 0xc00011e500, 0xc000139b00)
        D:/go1.13.8.windows-amd64/gopath/pkg/mod/github.com/amitbet/vnc2video@v0.0.0-20190616012314-9d50b9dab1d9/client.go:303 +0x2ab
created by github.com/amitbet/vnc2video.(*DefaultClientMessageHandler).Handle
        D:/go1.13.8.windows-amd64/gopath/pkg/mod/github.com/amitbet/vnc2video@v0.0.0-20190616012314-9d50b9dab1d9/client.go:285 +0x2b8
```

liuzheng commented:
I run in a goroutine, and when I create and distory frequently, it will show this panic.

I was not very sure which reason cause it, but this is my fix.

amitbet commented on encoding_util.go:
```diff
@@ -58,7 +58,7 @@ func (c *VncCanvas) Reset(rect *Rectangle) {
 
 func (c *VncCanvas) RemoveCursor() image.Image {
-	if c.Cursor == nil || c.CursorLocation == nil {
+	if c == nil || c.Cursor == nil || c.CursorLocation == nil {
```
Not having a canvas in your connection means that the Connect function in client.go did not succeed - you should check why...

Also, I don't see why this fix helps - since the next function to be called in client.go is canvas.PaintCursor:

canvas := c.(*ClientConn).Canvas
canvas.RemoveCursor()
parsedMsg, err := msg.Read(c)
canvas.PaintCursor()

gyuchang replied:
`c` could be `nil` if `.SetTargetImage(screenImage)` (in `example/client/main.go`) has not been called yet. because launching `ffmpeg` takes time and the message handlers are started before that in `vnc.Connect()`, `DefaultClientServerInitHandler` could be called before the renderers were given the pointer to the canvas.

I think ignoring missing canvas here could give the system just enough time to run `.SetTargetImage(screenImage)` so that by the time `PaintCursor()` is called, the renderers are ready.

See #14 for my proposed fix.

gyuchang committed:
Move Canvas and Renderer initialization to DefaultClientServerInitHandler
 - otherwise DefaultClientMessageHandler could get called before canvas and renderer are initialized

liuzheng committed:
revert(encoding_util.go): revert

liuzheng committed:
Merge branch 'fix-renderer-init' of https://github.com/gyuchang/vnc2video
